spmd/distributed.py

```diff
@@ -0,0 +1,94 @@
+import io
+from dataclasses import replace
+from typing import Any
+
+import torch
+import torch.distributed.distributed_c10d as c10d
+from torch.distributed import ProcessGroup
+from torch.distributed._shard.checkpoint.default_planner import (
+    DefaultLoadPlanner,
+    DefaultSavePlanner,
+)
+from torch.distributed._shard.checkpoint.metadata import (
+    STATE_DICT_TYPE,
+    Metadata,
+    MetadataIndex,
+)
+from torch.distributed._shard.checkpoint.planner import ReadItem, SavePlan
+
+
+def get_ranks(pg: ProcessGroup):
+    """
+    Return an array of global ranks for a given process group.
+    """
+    return [c10d._get_global_rank(pg, i) for i in range(pg.size())]
+
+
+class ProcessGroupAwareSavePlanner(DefaultSavePlanner):
+    """
+    SavePlanner class defining the protocol used by save_state_dict() to plan
+    a process group aware save plan.
+    """
+
+    def create_local_plan(self) -> SavePlan:
+        """
+        Each FSDP process group holds different data but use the same FQN in the state_dict.
+        This will modify the FQN in the save plan to be prefixed by a PG specific string.
+        """
+        save_plan = super().create_local_plan()
+
+        write_item = save_plan.items[0]
```

Comment:
We should perform the replacement on all ShardedTensor's in a state_dict that belong to a process group.

To see that behavior, you can use a model with multiple FSDP instances:
```
from torch.distributed.fsdp.wrap import (
    enable_wrap,
    wrap,
)
class SequentialModel(torch.nn.Module):
    def __init__(self, inner_module=None):
        super(SequentialModel, self).__init__()

        self.inner0 = wrap(NestedModule())
        self.inner1 = wrap(NestedModule())
        self.inner2 = wrap(NestedModule())
        self.inner3 = wrap(NestedModule())
        self.net1 = torch.nn.Linear(1000, 1000, device="meta")

    def reset_parameters(self):
        pass

with enable_wrap(wrapper_cls=FSDP, device_id=dist.get_rank(), process_group):
    model = wrap(SequentialModel())
```